File: CsBinBuff/BinBuffTest/Game.cs
using BinBuff;

namespace BinBuffTest
{
    class Game : BinBuff.ISerializable
    {
        

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Game) obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = width;
                hashCode = (hashCode * 397) ^ height;
                hashCode = (hashCode * 397) ^ (player != null ? player.GetHashCode() : 0);
                hashCode = (hashCode * 397) ^ (enemies != null ? enemies.GetHashCode() : 0);
                return hashCode;
            }
        }

        private int width, height;
        private Player player;
        private Player[] enemies;

        public Game()
        {
            width = 0;
            height = 0;
            player = null;
            enemies = null;
        }

        public Game(int width, int height, int numEnemies)
        {
            this.width = width;
            this.height = height;
            this.enemies = new Player[numEnemies];
            this.player = new Player(1);
            for (int i = 0; i < numEnemies; i++)
            {
                this.enemies[i] = new Player(i + 2);
            }
        }

        public static bool operator ==(Game g1, Game g2)
        {
            bool areEqual = g1.width == g2.width && g1.height == g2.height;
            if (!areEqual) return false;
            for (int i = 0; i < g1.enemies.Length; i++)
            {
                areEqual = areEqual && g1.enemies[i] == g2.enemies[i];
            }

            return areEqual && g1.player == g2.player;
        }
        protected bool Equals(Game other)
        {
            return this == other;
        }

        public static bool operator !=(Game g1, Game g2)
        {
            return !(g1 == g2);
        }

        public void Serialize(BinBuffer binBuffer)
        {
            binBuffer.Write(width, height);
            binBuffer.Write(player);
            binBuffer.Write((short)enemies.Length);
            binBuffer.Write(enemies);
        }

        public void Deserialize(BinBuffer binBuffer)
        {
            binBuffer.Read(out width);
            binBuffer.Read(out height);
            binBuffer.Read(out player);
            short numEnemies;
            binBuffer.Read(out numEnemies);
            enemies = new Player[numEnemies];
            binBuffer.Read(enemies, numEnemies);
        }
    }
}

File: CppBinBuff/test/Game.cpp
#include "Game.h"

Game::Game(int width, int height, int num_enemies) : width(width), height(height), num_enemies(num_enemies)
{
	this->player = new Player(1);
	this->enemies = new Player*[num_enemies];
	for (int i = 0; i < num_enemies; ++i)
	{
		this->enemies[i] = new Player(i + 2);
	}
}

Game::Game(const Game& other)
{
	this->player = new Player;
	*this->player = *other.player;
	this->enemies = new Player*[other.num_enemies];
	for (int i = 0; i < other.num_enemies; ++i)
	{
		this->enemies[i] = new Player(*other.enemies[i]);
	}
	this->num_enemies = other.num_enemies;
	this->height = other.height;
	this->width = other.width;
}

Game::~Game()
{
	delete this->player;
	for (int i = 0; i < this->num_enemies; ++i)
	{
		delete this->enemies[i];
	}
	delete[] this->enemies;
}

Game& Game::operator=(const Game& other)
{
	if (this->player == nullptr) this->player = new Player;
	*this->player = *other.player;
	for (int i = 0; i < this->num_enemies; ++i)
	{
		delete this->enemies[i];
	}
	delete[] this->enemies;
	this->enemies = new Player*[other.num_enemies];
	for (int i = 0; i < other.num_enemies; ++i)
	{
		this->enemies[i] = new Player;
		*this->enemies[i] = *other.enemies[i];
	}
	this->num_enemies = other.num_enemies;
	this->height = other.height;
	this->width = other.width;

	return *this;
}

Game& Game::operator=(Game&& other) noexcept
{
	delete this->player;
	for (int i = 0; i < this->num_enemies; ++i)
	{
		delete this->enemies[i];
	}
	delete[] this->enemies;

	this->player = other.player;
	this->enemies = other.enemies;
	this->height = other.height;
	this->num_enemies = other.num_enemies;
	this->width = other.width;

	other.player = nullptr;
	other.enemies = nullptr;
	other.height = 0;
	other.num_enemies = 0;
	other.width = 0;

	return *this;
}

bool Game::operator==(const Game& other) const
{
	bool areEqual = this->height == other.height && this->width == other.width;
	if (!areEqual) return false;
	for (int i = 0; i < this->num_enemies; ++i)
	{
		areEqual = areEqual && *this->enemies[i] == other.enemies[i];
	}
	return areEqual && *this->player == other.player;
}

void Game::serialize(bbf::Buffer& buffer) const
{
	buffer << this->width << this->height;
	buffer << *this->player;
	buffer << this->num_enemies;
	buffer.write(this->enemies, this->num_enemies);
}

void Game::deserialize(bbf::Buffer& buffer)
{
	buffer >> this->width >> this->height;
	this->player = new Player;
	buffer >> *this->player;
	buffer >> this->num_enemies;
	this->enemies = new Player*[this->num_enemies];
	buffer.read(this->enemies, this->num_enemies); 
}
